flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -146,7 +149,10 @@ private boolean add(BufferConsumer bufferConsumer, boolean finish, boolean inser
 			handleAddingBarrier(bufferConsumer, insertAsHead);
 			updateStatistics(bufferConsumer);
 			increaseBuffersInBacklog(bufferConsumer);
-			notifyDataAvailable = insertAsHead || finish || shouldNotifyDataAvailable();
+
+			// Notify only when we added first finished buffer.
+			notifyDataAvailable = mayNotifyDataAvailable() && getNumberOfFinishedBuffers() == 1;
+			dataAvailableNotified = dataAvailableNotified || notifyDataAvailable;
```

Comment:
it is not easy to identify `dataAvailableNotified` with `notifyDataAvailable` together. How about renaming `dataAvailableNotified` as `alreadyNotified` and `notifyDataAvailable` as `shouldNotify`, which might seem more clearly.

Reply:
We can remove `dataAvailableNotified = dataAvailableNotified || notifyDataAvailable` and migrate it into internal `#notifyDataAvailable()` to tag `dataAvailableNotified = true` instead. There are two benefits:

1. It is more easily understood after actually calling `#notifyDataAvailable`, then set it as true.
2. It can be reused for three below places, which i will point it out below.